packages/modal/src/Modal.tsx

```diff
@@ -0,0 +1,281 @@
+  font-weight: 200;
+  color: #464c4f;
+  position: relative;
+  float: right;
+  top: -18px;
+  right: -10px;
+  margin-top: -2px;
+  cursor: pointer;
+`;
+
+const titleStyle = css`
+  margin: 0;
+`;
+
+const EscapeKey = 27;
+
+interface ModalProps {
+  /**
+   * Content that will appear inside of the Modal component.
+   */
+  children: React.ReactNode;
+
+  /**
+   * Determines the active state of the modal
+   *
+   * default: `false`
+   */
+  active: boolean;
+
+  /**
+   * Specifies that the Modal content will appear portaled to the end of the DOM,
+   * rather than in the React DOM tree.
+   *
+   * default: `true`
+   */
+  usePortal?: boolean;
+
+  /**
+   * Specifies the size of the Modal.
+   *
+   * default: `normal`
+   */
+  size?: ModalSize;
+
+  /**
+   * Specifies the title of the Modal.
+   *
+   */
+  title?: string;
+
+  /**
+   * Callback to change the active state of the Modal.
+   *
+   */
+  setActive?: (
+    active: boolean,
+  ) => void | React.Dispatch<SetStateAction<boolean>>;
+
+  /**
+   * Callback invoked when Modal closes.
+   *
+   */
+  onRequestClose?: () => void;
+
+  /**
+   * className applied to root overlay div.
+   *
+   */
+  className?: string;
+}
+
+/**
+ * # Modal
+ *
+ *  Modals place content on top of main window.
+ *
+```
+<Modal
+  active
+  usePortal
+  size="large"
+  setActive={setActive}
+  title="My Modal"
+  onRequestClose={() => console.log('Modal is closing now!')}
+  >  
+  Modal content!
+</Modal>
+```
+ * @param props.active Boolean to describe whether or not Modal is active.
+ * @param props.usePortal Boolean to describe if content should be portaled to end of DOM, or appear in React DOM tree.
+ * @param props.size String to determine size of Modal. ['xxsmall', 'xsmall', 'small', 'normal', 'large', 'xlarge']
+ * @param props.setActive Callback to change the active state of Modal.
+ * @param props.children Content to appear inside of Modal container.
+ * @param props.title Title for the Modal, will appear inside header tags.
+ * @param props.onRequestClose Callback invoked when Modal is closed.
+ * @param props.className className applied to overlay div.
+ *
+ */
+function Modal({
+  active = false,
+  usePortal = true,
+  size = ModalSize.Normal,
+  setActive,
+  children,
+  title,
+  onRequestClose,
+  className,
+  ...rest
+}: ModalProps) {
+  const contentRef = useRef<HTMLDivElement>(null);
+
+  const handleClose = () => {
+    if (onRequestClose) {
+      onRequestClose();
+    }
+
+    if (setActive) {
+      setActive(false);
+    }
+  };
+
+  const handleEscape = useCallback((e: KeyboardEvent) => {
+    if (e.keyCode === EscapeKey) {
+      handleClose();
+    }
+  }, []);
+
+  const handleDocumentClick = (e: React.SyntheticEvent) => {
+    if (!contentRef.current) {
+      return;
+    }
+
+    if (!contentRef.current.contains(e.target as Node)) {
+      handleClose();
+    }
+  };
+
+  useEventListener('keydown', handleEscape, {
+    options: { once: true },
+  });
```

Comment:
`useModalState` should handle all of these event listeners, since that's the point of having it.

Reply:
Alternatively, maybe we don't need it?

```js
const { active, setActive } = useState(false)

return <Modal active={active} setActive={setActive}>{ children }</Modal>
```

In that case, `setActive` is optional, and handling the ref is much easier. Thinking about all this more, the `useModalState` hook use would have to look like:

```js
const { active, setActive, setRef } = useModalState(false)

return <Modal active={active} setActive={setActive} setRef={setRef}>{ children }</Modal>
```

Thoughts @hswolff? I'm starting to lean towards the former option here, now remembering that we would have to have a ref internally in the hook to prevent click events within the modal from closing it.

Reply:
I think if we were considering the `useModal` hook I proposed before, it would be a different story because the hook could set that ref itself without developer intervention:

```js
const { active, setActive, Modal } = useModalHook(false);

return <Modal>{ children }</Modal>;
```